url3010dbde70_nbartlett_Sequence_Memoizer_tgz-pJ8-edu_columbia_stat_wood_pub_sequencememoizer_IntSequenceMemoizerJ8/src/edu/columbia/stat/wood/pub/sequencememoizer/util/IntRestaurant.java

```diff
@@ -0,0 +1,478 @@
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.io.Serializable;
+
+/**
+ * Node object used in int sequence memoizer.
+ * @author nicholasbartlett
+ */
+public class IntRestaurant extends IntMap<IntRestaurant> implements Serializable{
+
+    static final long serialVersionUID = 1;
+
+    /**
+     * Types observed this node.
+     */
+    public int[] types;
+
+    /**
+     * Counts of number of types and the number of tables (using the Chinese Restaurant metaphor)
+     * at which they are sitting.
+     */
+    public int[] customersAndTables;
+
+    /**
+     * Total number of observations seen in this node.
+     */
+    public int customers;
+    
+    /**
+     * Total number of tables in this node.
+     */
+    public int tables;
+
+    /**
+     * Offset between edge node start and the start of the edge label.
+     */
+    public int edgeStart;
+
+    /**
+     * Length of the edge.
+     */
+    public int edgeLength;
+
+    /**
+     * Number of leaf nodes below and including this node.
+     */
+    public int numLeafNodesAtOrBelow;
+
+    /**
+     * Parent node.
+     */
+    public IntRestaurant parent;
+
+    /**
+     * Node of linked list where this edge starts.
+     */
+    public IntSeqNode edgeNode;
+
+    public IntRestaurant(IntRestaurant parent, int edgeStart,  int edgeLength, IntSeqNode edgeNode,  int numLeafNodesAtOrBelow, MutableLong restCount) {
+        this.parent = parent;
+        this.edgeStart = edgeStart;
+        this.edgeLength = edgeLength;
+        this.edgeNode = edgeNode;
+        this.numLeafNodesAtOrBelow = numLeafNodesAtOrBelow;
+
+        if(edgeNode != null){
+            edgeNode.add(this);
+        }
+        customers = 0;
+        tables = 0;
+        restCount.increment();
+    }
+
+    public void setTableConfig(int[] types, int[] customersAndTables, int customers, int tables) {
+        this.types = types;
+        this.customersAndTables = customersAndTables;
+        this.customers = customers;
+        this.tables = tables;
+    }
+
+    public double getPP(int type, double p, double discount, SeatReturn sr) {
+        int index, tc, tt, tci, tti;
+
+
+        if(type > types[types.length -1]){
+            tt = 0;
+            tc = 0;
+        } else {
+            index = getIndex(type);
+            if(types[index] == type){
+                tci = 2 * index;
+                tti = tci + 1;
+
+                tc = customersAndTables[tci];
+                tt = customersAndTables[tti];
+            } else {
+                tt = 0;
+                tc = 0;
+            }
+        }
+        
+        sr.set(false, tt, customers, tables);
+        p -= ((double) tc - (double) tt * discount) / (double) customers;
+
+        return p * (double) customers / ((double) tables * discount);
+    }
+
+    public void deleteCustomers(int nDelete, double discount) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 13. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 157-175 into a separate method.